clouddriver-azure/src/main/groovy/com/netflix/spinnaker/clouddriver/azure/client/AzureNetworkClient.groovy

```diff
@@ -351,6 +362,213 @@ class AzureNetworkClient extends AzureBaseClient {
+      List<int[]> usedPortList = new ArrayList<>()
+      for(LoadBalancerInboundNatPool pool : loadBalancer.inboundNatPools().values()) {
+        int[] range = new int[2]
+        range[0] = pool.frontendPortRangeStart()
+        range[1] = pool.frontendPortRangeEnd()
+        usedPortList.add(range)
+      }
+      usedPortList.sort(true, new Comparator<int[]>() {
+        @Override
+        int compare(int[] o1, int[] o2) {
+          return o1[0] - o2[0]
+        }
+      })
+
+      if (loadBalancer.inboundNatPools()?.containsKey(serverGroupName)) {
+        return loadBalancer.inboundNatPools().get(serverGroupName).inner().id()
+      }
+      int portStart = findUnusedPortsRange(usedPortList, NAT_POOL_PORT_START, NAT_POOL_PORT_END, NAT_POOL_PORT_NUMBER_PER_POOL)
+      if(portStart == -1) {
+        throw new RuntimeException("Load balancer ${loadBalancerName} does not have unused port between ${NAT_POOL_PORT_START} and ${NAT_POOL_PORT_END} with length ${NAT_POOL_PORT_NUMBER_PER_POOL}")
+      }
+
+      // The purpose of the following code is to create an NAT pool in an existing Azure Load Balancer
+      // However Azure Java SDK doesn't provide a way to do this
+      // Use reflection to modify the backend of load balancer
+      LoadBalancerInboundNatPool.UpdateDefinitionStages.WithAttach< LoadBalancer.Update> update = loadBalancer.update()
+        .defineInboundNatPool(serverGroupName)
+        .withProtocol(TransportProtocol.TCP)
+
+      try {
+        Method setRangeMethod = update.getClass().getMethod("fromFrontendPortRange", int.class, int.class)
+        setRangeMethod.setAccessible(true)
+        setRangeMethod.invoke(update, portStart, portStart + NAT_POOL_PORT_NUMBER_PER_POOL - 1)
+
+        Method setBackendPortMethod = update.getClass().getMethod("fromFrontend", String.class)
+        setBackendPortMethod.setAccessible(true)
+        setBackendPortMethod.invoke(update, frontEndName)
+
+        Method setFrontendMethod = update.getClass().getMethod("toBackendPort", int.class)
+        setFrontendMethod.setAccessible(true)
+        setFrontendMethod.invoke(update, 22)
+      } catch (NoSuchMethodException e) {
+        log.error("Failed to use reflection to create NAT pool in Load Balancer, detail: {}", e.getMessage())
+        return null
+      } catch (IllegalAccessException e) {
+        log.error("Failed to use reflection to create NAT pool in Load Balancer, detail: {}", e.getMessage())
+        return null
+      } catch (InvocationTargetException e) {
+        log.error("Failed to use reflection to create NAT pool in Load Balancer, detail: {}", e.getMessage())
+        return null
+      }
+
+      update.attach().apply()
+      return loadBalancer.inboundNatPools().get(serverGroupName).inner().id()
+
+    } else {
+      throw new RuntimeException("Load balancer ${loadBalancerName} not found in resource group ${resourceGroupName}")
+    }
+
+    null
+  }
+
+  // Find unused port range
```

Comment:
Probably don't need this comment, as it's just the name of the method.

Reply:
Refine the comment